File: 1.py
# def dfs(graph, start):
#     visited = set()
#     traversal_order = []

#     def dfs_helper(node):
#         visited.add(node)
#         traversal_order.append(node)
#         for neighbor in graph[node]:
#             if neighbor not in visited:
#                 dfs_helper(neighbor)

#     dfs_helper(start)
#     return traversal_order

# # Example usage:
# graph = {
#     'A': ['B', 'C'],
#     'B': ['D', 'E'],
#     'C': ['F'],
#     'D': [],
#     'E': ['F'],
#     'F': []
# }

# print("Depth-First Traversal:", dfs(graph, 'A'))



from collections import deque

def bfs(graph, start):
    visited = set()
    traversal_order = []
    queue = deque([start])

    while queue:
        node = queue.popleft()
        if node not in visited:
            visited.add(node)
            traversal_order.append(node)
            for neighbor in graph[node]:
                if neighbor not in visited:
                    queue.append(neighbor)

    return traversal_order

# Example usage:
graph = {
    'A': ['B', 'C'],
    'B': ['D', 'E'],
    'C': ['F'],
    'D': [],
    'E': ['F'],
    'F': []
}

print("Breadth-First Traversal:", bfs(graph, 'A'))



# def recursive_bfs(graph, start):
#     visited = set()
#     traversal_order = []

#     def bfs_helper(node):
#         visited.add(node)
#         traversal_order.append(node)
#         for neighbor in graph[node]:
#             if neighbor not in visited:
#                 bfs_helper(neighbor)

#     bfs_helper(start)
#     return traversal_order

# # Example usage:
# graph = {
#     'A': ['B', 'C'],
#     'B': ['D', 'E'],
#     'C': ['F'],
#     'D': [],
#     'E': ['F'],
#     'F': []
# }

# print("Recursive Breadth-First Traversal:", recursive_bfs(graph, 'A'))


##########################################################################################################################
switch case driven code
from collections import deque

def dfs(graph, start):
    visited = set()
    stack = [start]
    while stack:
        vertex = stack.pop()
        if vertex not in visited:
            print(vertex, end=" ")
            visited.add(vertex)
            stack.extend(graph[vertex] - visited)

def bfs(graph, start):
    visited = set()
    queue = deque([start])
    visited.add(start)
    while queue:
        vertex = queue.popleft()
        print(vertex, end=" ")
        for neighbor in graph[vertex] - visited:
            visited.add(neighbor)
            queue.append(neighbor)

def switch_case(choice, graph, start):
    switcher = {
        'dfs': dfs,
        'bfs': bfs
    }
    # Get the function from switcher dictionary
    func = switcher.get(choice.lower())
    # Execute the function
    if func:
        func(graph, start)
    else:
        print("Invalid choice")

# Example usage
graph = {
    'A': {'B', 'C'},
    'B': {'A', 'D', 'E'},
    'C': {'A', 'F'},
    'D': {'B'},
    'E': {'B', 'F'},
    'F': {'C', 'E'}
}
start_vertex = 'A'
choice = input("Enter 'dfs' for Depth-First Search or 'bfs' for Breadth-First Search: ")
switch_case(choice, graph, start_vertex)

File: 1.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Graph {
    int V;
    vector<vector<int>> adjList;

public:
    Graph(int V) {
        this->V = V;
        adjList.resize(V);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    void DFSUtil(int v, bool visited[]) {
        visited[v] = true;
        cout << v << " ";
        for (int i = 0; i < adjList[v].size(); i++) {
            if (!visited[adjList[v][i]])
                DFSUtil(adjList[v][i], visited);
        }
    }

    void DFS(int start) {
        bool visited[V];
        for (int i = 0; i < V; i++)
            visited[i] = false;

        DFSUtil(start, visited);
    }
};

int main() {
    Graph g(5);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 4);

    cout << "Depth-first traversal for the given graph starting from node 0: ";
    g.DFS(0);

    return 0;
}



#include <iostream>
#include <vector>
#include <queue>
#include <list>

using namespace std;

class Graph {
    int V;
    vector<vector<int>> adjList;

public:
    Graph(int V) {
        this->V = V;
        adjList.resize(V);
    }

    void addEdge(int u, int v) {
        adjList[u].push_back(v);
        adjList[v].push_back(u);
    }

    void BFS(int start) {
        vector<bool> visited(V, false);
        queue<int> q;

        visited[start] = true;
        q.push(start);

        while (!q.empty()) {
            int node = q.front();
            q.pop();
            cout << node << " ";

            for (int i = 0; i < adjList[node].size(); i++) {
                if (!visited[adjList[node][i]]) {
                    visited[adjList[node][i]] = true;
                    q.push(adjList[node][i]);
                }
            }
        }
    }
};

int main() {
    Graph g(5);
    g.addEdge(0, 1);
    g.addEdge(0, 2);
    g.addEdge(1, 3);
    g.addEdge(2, 4);

    cout << "Breadth-first traversal for the given graph starting from node 0: ";
    g.BFS(0);

    return 0;
}